File: livedata-lesson/src/main/java/com/acme/recipes/data/util/RealmLiveData.java
package com.acme.recipes.data.util;

import android.arch.lifecycle.LiveData;

import javax.annotation.Nonnull;

import io.realm.RealmChangeListener;
import io.realm.RealmModel;
import io.realm.RealmObject;

public class RealmLiveData <T extends RealmModel> extends LiveData<T> {

    private T model;

    private RealmChangeListener<T> listener;

    public RealmLiveData(T model) {
        this.model = model;

        listener =new RealmChangeListener<T>() {
            @Override
            public void onChange(@Nonnull T update) {
                setValue(update);
            }
        };

    }

    @Override
    protected void onActive() {
        super.onActive();
        RealmObject.addChangeListener(model, listener);
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        RealmObject.removeChangeListener(model, listener);
    }

}

File: app/src/main/java/com/acme/recipes/database/util/RealmLiveData.kt
package com.acme.recipes.database.util

import android.arch.lifecycle.LiveData

import io.realm.RealmChangeListener
import io.realm.RealmModel
import io.realm.RealmObject

class RealmLiveData<T : RealmModel>(private val model: T) : LiveData<T>() {

    private val listener = RealmChangeListener<T> { update -> value = update }

    override fun onActive() {
        super.onActive()
        RealmObject.addChangeListener(model, listener)
    }

    override fun onInactive() {
        super.onInactive()
        RealmObject.removeChangeListener(model, listener)
    }

}